From: IrfanMohammad <irfan@graphiant.com>
Date: Wed, 26 Jul 2023 18:17:37 +0000
Subject: Add change_sub with minimal locks

- do not lock mods longer than necessary for adding enabled change_sub

    Addition of change_sub happens in 5 important steps w.r.t locks
    1. change_sub write lock- This will prevent any new events from
       being published for this module
    2. modules lock, invoke enabled callback() and unlock
    3. add sub to ext_shm and downgrade change_sub to read lock- Any
       writers can now notify subscribers, but this subscription is not
       ready yet.
    4. acquire subs_lock write mode, add subscr to subscription
       structure to make it visible to handler thread
    5. release write subs_lock, then change_sub read lock

    Previously, we held module locks beyond Step 2.
    This was to prevent losing events after Step 2 and before Step 5.

    Since the fix for ENG-3166 and ENG-3839, We always check for any
    missed events during Step 4, and notify the handler thread.

    Because of this, holding onto modules lock is no longer necessary.

cherry-picked from 0067-Locking-improvements-for-subscriptions.patch
---
 src/shm_ext.c | 50 ++++++++++++++++++++++++++++++++---
 src/shm_ext.h |  7 ++---
 src/subscr.c  | 15 ++++++++---
 src/subscr.h  |  2 +-
 src/sysrepo.c | 83 +++++++++++++++++++++++++++++++++++++----------------------
 5 files changed, 114 insertions(+), 43 deletions(-)

diff --git a/src/shm_ext.c b/src/shm_ext.c
index 8e133e8..798ce73 100644
--- a/src/shm_ext.c
+++ b/src/shm_ext.c
@@ -555,14 +555,17 @@ sr_shmext_change_sub_remove_dead(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_data
 }
 
 sr_error_info_t *
-sr_shmext_change_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_datastore_t ds, uint32_t sub_id, const char *xpath,
-        uint32_t priority, int sub_opts, uint32_t evpipe_num)
+sr_shmext_change_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_lock_mode_t *has_lock,
+        sr_datastore_t ds, uint32_t sub_id, const char *xpath, uint32_t priority, int sub_opts, uint32_t evpipe_num)
 {
     sr_error_info_t *err_info = NULL, *tmp_err;
     off_t xpath_off;
     sr_mod_change_sub_t *shm_sub;
     uint32_t i;
 
+    /* kept for possible future use */
+    assert(*has_lock == SR_LOCK_WRITE);
+
     /* EXT WRITE LOCK */
     if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_WRITE, 1, __func__))) {
         goto cleanup;
@@ -628,6 +631,17 @@ sr_shmext_change_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_datastore_t
 cleanup_unlock:
     /* EXT WRITE UNLOCK */
     sr_shmext_conn_remap_unlock(conn, SR_LOCK_WRITE, 1, __func__);
+    /*
+     * Downgrade the write lock to a read upgrade lock, we will unlock after everything is finished
+     * read upgrade lock needed, because we may need to delete the sub if next steps fail
+     * */
+    if (err_info) {
+        goto cleanup;
+    }
+    if (!(err_info = sr_rwrelock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ_UPGR, conn->cid,
+            __func__, NULL, NULL))) {
+        *has_lock = SR_LOCK_READ_UPGR;
+    }
 
 cleanup:
     return err_info;
@@ -738,15 +752,31 @@ cleanup:
 }
 
 sr_error_info_t *
-sr_shmext_change_sub_del(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_datastore_t ds, uint32_t sub_id)
+sr_shmext_change_sub_del(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_lock_mode_t *has_lock,
+        sr_datastore_t ds, uint32_t sub_id)
 {
     sr_error_info_t *err_info = NULL;
     sr_mod_change_sub_t *shm_sub;
     uint32_t i;
 
+    assert((*has_lock == SR_LOCK_NONE) || (*has_lock == SR_LOCK_WRITE || *has_lock == SR_LOCK_READ_UPGR));
+
+    if (*has_lock == SR_LOCK_NONE) {
+        /* CHANGE SUB WRITE LOCK */
+        if ((err_info = sr_rwlock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_WRITE, conn->cid,
+                __func__, NULL, NULL))) {
+            return err_info;
+        }
+    } else if (*has_lock == SR_LOCK_READ_UPGR) {
+        if ((err_info = sr_rwrelock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_WRITE, conn->cid,
+                __func__, NULL, NULL))) {
+            return err_info;
+        }
+    }
+
     /* EXT READ LOCK */
     if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_READ, 1, __func__))) {
-        return err_info;
+        goto cleanup_chsub_unlock;
     }
 
     /* find the subscription */
@@ -770,6 +800,18 @@ cleanup_unlock:
     /* EXT READ UNLOCK */
     sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 1, __func__);
 
+cleanup_chsub_unlock:
+    if (*has_lock == SR_LOCK_NONE) {
+        /* CHANGE SUB WRITE UNLOCK */
+        sr_rwunlock(&shm_mod->change_sub[ds].lock, 0, SR_LOCK_WRITE, conn->cid, __func__);
+    } else if (*has_lock == SR_LOCK_READ_UPGR) {
+        /* Downgrade to a read lock */
+        if ((err_info = sr_rwrelock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid,
+                __func__, NULL, NULL))) {
+            return err_info;
+        }
+        *has_lock = SR_LOCK_READ;
+    }
     return err_info;
 }
 
diff --git a/src/shm_ext.h b/src/shm_ext.h
index f21ba48..a4979d1 100644
--- a/src/shm_ext.h
+++ b/src/shm_ext.h
@@ -72,8 +72,8 @@ void sr_shmext_print(sr_mod_shm_t *mod_shm, sr_shm_t *shm_ext);
  * @param[in] evpipe_num Subscription event pipe number.
  * @return err_info, NULL on success.
  */
-sr_error_info_t *sr_shmext_change_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_datastore_t ds, uint32_t sub_id,
-        const char *xpath, uint32_t priority, int sub_opts, uint32_t evpipe_num);
+sr_error_info_t *sr_shmext_change_sub_add(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_lock_mode_t *chsub_lock_mode,
+        sr_datastore_t ds, uint32_t sub_id, const char *xpath, uint32_t priority, int sub_opts, uint32_t evpipe_num);
 
 /**
  * @brief Modify existing main SHM module change subscription.
@@ -98,7 +98,8 @@ sr_error_info_t *sr_shmext_change_sub_modify(sr_conn_ctx_t *conn, sr_mod_t *shm_
  * @param[in] sub_id Unique sub ID.
  * @return err_info, NULL on success.
  */
-sr_error_info_t *sr_shmext_change_sub_del(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_datastore_t ds, uint32_t sub_id);
+sr_error_info_t *sr_shmext_change_sub_del(sr_conn_ctx_t *conn, sr_mod_t *shm_mod, sr_lock_mode_t *chsub_lock_mode,
+        sr_datastore_t ds, uint32_t sub_id);
 
 /**
  * @brief Remove main SHM module change subscription with param-based cleanup.
diff --git a/src/subscr.c b/src/subscr.c
index 3dd19fe..c618bb6 100644
--- a/src/subscr.c
+++ b/src/subscr.c
@@ -39,13 +39,14 @@
 sr_error_info_t *
 sr_subscr_change_sub_add(sr_subscription_ctx_t *subscr, uint32_t sub_id, sr_session_ctx_t *sess, const char *mod_name,
         const char *xpath, sr_module_change_cb change_cb, void *private_data, uint32_t priority,
-        sr_subscr_options_t sub_opts, sr_lock_mode_t has_subs_lock)
+        sr_subscr_options_t sub_opts, sr_lock_mode_t has_subs_lock, int *is_event_pending)
 {
     sr_error_info_t *err_info = NULL;
     struct modsub_change_s *change_sub = NULL;
     uint32_t i;
     void *mem[4] = {NULL};
     int new_sub = 0;
+    sr_multi_sub_shm_t *multi_sub_shm;
 
     /* just to prevent problems in future changes */
     assert(has_subs_lock == SR_LOCK_WRITE);
@@ -110,6 +111,9 @@ sr_subscr_change_sub_add(sr_subscription_ctx_t *subscr, uint32_t sub_id, sr_sess
     /* new subscription */
     subscr->last_sub_id = sub_id;
 
+    multi_sub_shm = (sr_multi_sub_shm_t *)(change_sub->sub_shm.addr);
+    *is_event_pending = ATOMIC_LOAD_RELAXED(multi_sub_shm->subscriber_count) != 0;
+
     return NULL;
 
 error:
@@ -937,6 +941,7 @@ sr_change_sub_del(sr_subscription_ctx_t *subscr, uint32_t idx1, uint32_t idx2, s
     sr_datastore_t ds;
     sr_lock_mode_t subs_lock = has_subs_lock;
     uint32_t sub_id;
+    sr_lock_mode_t chsub_lock_mode = SR_LOCK_NONE;
 
     assert(has_subs_lock == SR_LOCK_READ);
     (void)has_subs_lock;
@@ -961,6 +966,7 @@ sr_change_sub_del(sr_subscription_ctx_t *subscr, uint32_t idx1, uint32_t idx2, s
             subscr->conn->cid, __func__, NULL, NULL))) {
         goto cleanup;
     }
+    chsub_lock_mode = SR_LOCK_WRITE;
 
     /* SUBS WRITE LOCK */
     if ((err_info = sr_rwlock(&subscr->subs_lock, SR_SUBSCR_LOCK_TIMEOUT, SR_LOCK_WRITE, subscr->conn->cid,
@@ -980,7 +986,7 @@ sr_change_sub_del(sr_subscription_ctx_t *subscr, uint32_t idx1, uint32_t idx2, s
 
     /* properly remove the subscription from ext SHM, with separate specific SHM segment if no longer needed while
      * holding CHANGE SUB lock to prevent data races and processing events after the subscription is removed from SHM */
-    if ((err_info = sr_shmext_change_sub_del(subscr->conn, shm_mod, ds, sub_id))) {
+    if ((err_info = sr_shmext_change_sub_del(subscr->conn, shm_mod, &chsub_lock_mode, ds, sub_id))) {
         goto cleanup_unlock;
     }
 
@@ -989,8 +995,9 @@ sr_change_sub_del(sr_subscription_ctx_t *subscr, uint32_t idx1, uint32_t idx2, s
 
 cleanup_unlock:
     /* CHANGE SUB WRITE UNLOCK */
-    sr_rwunlock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_WRITE, subscr->conn->cid, __func__);
-
+    if (chsub_lock_mode != SR_LOCK_NONE) {
+        sr_rwunlock(&shm_mod->change_sub[ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, chsub_lock_mode, subscr->conn->cid, __func__);
+    }
 cleanup:
     if (subs_lock != has_subs_lock) {
         if (subs_lock == SR_LOCK_NONE) {
diff --git a/src/subscr.h b/src/subscr.h
index 9d5f2b1..710287e 100644
--- a/src/subscr.h
+++ b/src/subscr.h
@@ -40,7 +40,7 @@
  */
 sr_error_info_t *sr_subscr_change_sub_add(sr_subscription_ctx_t *subscr, uint32_t sub_id, sr_session_ctx_t *sess,
         const char *mod_name, const char *xpath, sr_module_change_cb change_cb, void *private_data, uint32_t priority,
-        sr_subscr_options_t sub_opts, sr_lock_mode_t has_subs_lock);
+        sr_subscr_options_t sub_opts, sr_lock_mode_t has_subs_lock, int *is_event_pending);
 
 /**
  * @brief Delete a change subscription from a subscription structure.
diff --git a/src/sysrepo.c b/src/sysrepo.c
index e0a78c9..a9736bc 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -5669,7 +5669,6 @@ cleanup:
  * @brief Perform enabled event on a subscription.
  *
  * @param[in] session Session to use.
- * @param[in,out] mod_info Empty mod info structure to use. If any modules were locked, they are kept that way.
  * @param[in] ly_mod Specific module.
  * @param[in] xpath Optional subscription xpath.
  * @param[in] callback Callback to call.
@@ -5679,7 +5678,7 @@ cleanup:
  * @return err_info, NULL on success.
  */
 static sr_error_info_t *
-sr_module_change_subscribe_enable(sr_session_ctx_t *session, struct sr_mod_info_s *mod_info,
+sr_module_change_subscribe_enable(sr_session_ctx_t *session,
         const struct lys_module *ly_mod, const char *xpath, sr_module_change_cb callback, void *private_data,
         uint32_t sub_id, int opts)
 {
@@ -5688,6 +5687,9 @@ sr_module_change_subscribe_enable(sr_session_ctx_t *session, struct sr_mod_info_
     sr_session_ctx_t *ev_sess = NULL;
     sr_error_t err_code;
 
+    struct sr_mod_info_s mod_info_, *mod_info = &mod_info_;
+    SR_MODINFO_INIT(mod_info_, session->conn, session->ds, session->ds == SR_DS_OPERATIONAL ? SR_DS_RUNNING : session->ds, 0);
+
     /* create mod_info structure with this module only, do not use cache to allow reading data in the callback
      * (avoid dead-lock) */
     if ((err_info = sr_modinfo_add(ly_mod, NULL, 0, 0, mod_info))) {
@@ -5759,6 +5761,10 @@ sr_module_change_subscribe_enable(sr_session_ctx_t *session, struct sr_mod_info_
     callback(ev_sess, sub_id, ly_mod->name, xpath, sr_ev2api(ev_sess->ev), 0, private_data);
 
 cleanup:
+    /* Unlock mod_info we no longer need it, because we will check for missed events later anyway */
+    sr_shmmod_modinfo_unlock(mod_info);
+    sr_modinfo_erase(mod_info);
+
     sr_session_stop(ev_sess);
     lyd_free_all(enabled_data);
     return err_info;
@@ -5848,11 +5854,13 @@ sr_module_change_subscribe(sr_session_ctx_t *session, const char *module_name, c
 {
     sr_error_info_t *err_info = NULL, *tmp_err;
     const struct lys_module *ly_mod;
-    struct sr_mod_info_s mod_info;
     sr_conn_ctx_t *conn;
     uint32_t sub_id;
     sr_subscr_options_t sub_opts;
     sr_mod_t *shm_mod;
+    sr_lock_mode_t chsub_lock_mode = SR_LOCK_NONE;
+    int subs_lock_held = 0;
+    int is_event_pending = 0;
 
     SR_CHECK_ARG_APIRET(!session || !SR_IS_STANDARD_DS(session->ds) || SR_IS_EVENT_SESS(session) || !module_name ||
             !callback || !subscription, session, err_info);
@@ -5862,7 +5870,8 @@ sr_module_change_subscribe(sr_session_ctx_t *session, const char *module_name, c
         xpath = NULL;
     }
 
-    SR_MODINFO_INIT(mod_info, session->conn, session->ds, session->ds == SR_DS_OPERATIONAL ? SR_DS_RUNNING : session->ds, 0);
+    SR_LOG_INF("%s: mod %s xpath %s ds %d priority %u opts %d", __func__, module_name, xpath,
+                session->ds, priority, opts);
 
     conn = session->conn;
     /* only these options are relevant outside this function and will be stored */
@@ -5916,30 +5925,37 @@ sr_module_change_subscribe(sr_session_ctx_t *session, const char *module_name, c
             conn->cid, __func__, NULL, NULL))) {
         goto cleanup;
     }
+    chsub_lock_mode = SR_LOCK_WRITE;
 
     if (opts & SR_SUBSCR_ENABLED) {
-        /* call the callback with the current configuration, keep any used modules locked in mod_info */
-        if ((err_info = sr_module_change_subscribe_enable(session, &mod_info, ly_mod, xpath, callback, private_data,
+        /* call the callback with the current configuration */
+        if ((err_info = sr_module_change_subscribe_enable(session, ly_mod, xpath, callback, private_data,
                 sub_id, opts))) {
-            goto cleanup_unlock1;
+            goto cleanup;
         }
     }
 
-    /* SUBS WRITE LOCK */
-    if ((err_info = sr_rwlock(&(*subscription)->subs_lock, SR_SUBSCR_LOCK_TIMEOUT, SR_LOCK_WRITE, conn->cid,
-            __func__, NULL, NULL))) {
-        goto cleanup_unlock1;
-    }
+    /* Do not need subs_lock here, since we are not modifying it yet.
+     * If an event is published after `sr_shmext_change_subscription_add` but before WRITE `subs_lock`
+     * We will miss it, So check for it in `sr_sub_change_add`
+     */
 
     /* add module subscription into ext SHM and create separate specific SHM segment */
-    if ((err_info = sr_shmext_change_sub_add(conn, shm_mod, session->ds, sub_id, xpath, priority, sub_opts,
+    if ((err_info = sr_shmext_change_sub_add(conn, shm_mod, &chsub_lock_mode, session->ds, sub_id, xpath, priority, sub_opts,
             (*subscription)->evpipe_num))) {
-        goto cleanup_unlock2;
+        goto cleanup;
+    }
+
+    /* SUBS WRITE LOCK */
+    if ((err_info = sr_rwlock(&(*subscription)->subs_lock, SR_SUBSCR_LOCK_TIMEOUT, SR_LOCK_WRITE, conn->cid,
+            __func__, NULL, NULL))) {
+        goto error1;
     }
+    subs_lock_held = 1;
 
     /* add subscription into structure */
     if ((err_info = sr_subscr_change_sub_add(*subscription, sub_id, session, module_name, xpath, callback, private_data,
-            priority, sub_opts, SR_LOCK_WRITE))) {
+            priority, sub_opts, SR_LOCK_WRITE, &is_event_pending))) {
         goto error1;
     }
 
@@ -5948,34 +5964,39 @@ sr_module_change_subscribe(sr_session_ctx_t *session, const char *module_name, c
             *subscription))) {
         goto error2;
     }
+    /* Notify the evpipe in case we missed an event */
+    if (is_event_pending) {
+        sr_shmsub_notify_evpipe((*subscription)->evpipe_num);
+    }
 
-    goto cleanup_unlock2;
+    goto cleanup;
 
 error2:
     sr_subscr_change_sub_del(*subscription, sub_id);
 
 error1:
-    if ((tmp_err = sr_shmext_change_sub_del(conn, shm_mod, session->ds, sub_id))) {
-        sr_errinfo_merge(&err_info, tmp_err);
+    if (subs_lock_held) {
+        /* SUBS WRITE UNLOCK */
+        sr_rwunlock(&(*subscription)->subs_lock, 0, SR_LOCK_WRITE, conn->cid, __func__);
+        subs_lock_held = 0;
     }
 
-cleanup_unlock2:
-    /* SUBS WRITE UNLOCK */
-    sr_rwunlock(&(*subscription)->subs_lock, 0, SR_LOCK_WRITE, conn->cid, __func__);
-
-cleanup_unlock1:
-    /* CHANGE SUB UNLOCK */
-    sr_rwunlock(&shm_mod->change_sub[session->ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_WRITE, conn->cid, __func__);
+    if ((tmp_err = sr_shmext_change_sub_del(conn, shm_mod, &chsub_lock_mode, session->ds, sub_id))) {
+        sr_errinfo_merge(&err_info, tmp_err);
+    }
 
-    /* if there are any modules, unlock them after the enabled event was handled and the subscription was added
-     * to avoid losing any changes */
+cleanup:
+    if (subs_lock_held) {
+        /* SUBS WRITE UNLOCK */
+        sr_rwunlock(&(*subscription)->subs_lock, 0, SR_LOCK_WRITE, conn->cid, __func__);
+    }
 
-    /* MODULES UNLOCK */
-    sr_shmmod_modinfo_unlock(&mod_info);
+    if (chsub_lock_mode != SR_LOCK_NONE) {
+        /* CHANGE SUB UNLOCK */
+        sr_rwunlock(&shm_mod->change_sub[session->ds].lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, chsub_lock_mode, conn->cid, __func__);
+    }
 
-    sr_modinfo_erase(&mod_info);
 
-cleanup:
     /* CONTEXT UNLOCK */
     sr_lycc_unlock(conn, SR_LOCK_READ, 0, __func__);
     return sr_api_ret(session, err_info);
